book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,305 @@
+| `utxo_by_outpoint`   | `OutPoint`            | `TransparentOutput`                 |
+| `sprout_nullifiers`  | `sprout::Nullifier`   | `()`                                |
+| `sapling_nullifiers` | `sapling::Nullifier`  | `()`                                |
+
+Zcash structures are encoded using `ZcashSerialize`/`ZcashDeserialize`.
+
+## Request / Response API
+[request-response]: #request-response
+
+The state API is provided by a pair of `Request`/`Response` enums. Each
+`Request` variant corresponds to particular `Response` variants, and it's
+fine (and encouraged) for caller code to unwrap the expected variants with
+`unreachable!` on the unexpected variants. This is slightly inconvenient but
+it means that we have a unified state interface with unified backpressure.
+
+This API includes both write and read calls. Spotting `Commit` requests in
+code review should not be a problem, but in the future, if we need to
+restrict access to write calls, we could implement a wrapper service that
+rejects these, and export "read" and "write" frontends to the same inner service.
+
+### `Request::CommitBlock(Arc<Block>)`
+[request-commit-block]: #request-commit-block
+
+Performs contextual validation of the given block, committing it to the state
+if successful. Returns `Response::Added(BlockHeaderHash)` with the hash of
+the newly committed block or an error.
+
+If the parent block is not committed, add the block to an internal queue for
+future processing.
+
+Otherwise, attempt to perform contextual validation checks and the commit the
+given block to the state. The exact list of contextual validation checks will
+be specified in a later RFC. If contextual validation checks succeed, the new
+block is added to one of the in-memory chains. If the resulting chain is
+longer than 100 blocks, the oldest block is now past the reorg limit, so we
+remove it from all in-memory chains and commit it to sled as described below
+in `CommitFinalizedBlock`.
+
+Finally, process any queued children of the newly committed block the same way.
+
+### `Request::CommitFinalizedBlock(Arc<Block>)`
+[request-commit-finalized-block]: #request-finalized-block
+
+Commits a finalized block to the sled state, skipping contextual validation.
+The block's parent must be the current sled tip. This is exposed for use in
+checkpointing, which produces in-order finalized blocks. Returns
+`Response::Added(BlockHeaderHash)` with the hash of the committed block if
+successful.
+
+This should be implemented as a wrapper around a function also called by
+[`Request::CommitBlock`](#request-commit-block), which should:
+
+1. Obtain the highest entry of `hash_by_height` as `(old_height, old_tip)`.
+Check that `block`'s parent hash is `old_tip` and its height is
+`old_height+1`, or panic.  This check is performed as defense-in-depth
+to prevent database corruption, but it is the caller's responsibility to
+commit finalized blocks in order.
```

Comment:
> it is the caller's responsibility to commit finalized blocks in order

In #969, we need to commit blocks in the future returned by the checkpoint verifier. To ensure that these futures execute in order, each checkpoint verifier future will need to await the commitment of the previous checkpoint range.

So let's add an AwaitBlock request to this RFC, which returns a future that completes when a block is committed to the state?

Reply:
As another alternative, we could re-use the CommitBlock buffer, but skip most of the checks.

Reply:
Or we could pass an un-buffered state to the checkpoint verifier.

Which is completely safe, because the buffer is only required once we're using the full block verifier, and its first block gets queued until the checkpoint verifier finishes.

Reply:
We decided to go with the buffer in the state service here.

Reply:
Noted this change in 6b8a5d7b